src/Markup/Avalonia.Markup.Xaml/XamlIl/CompilerExtensions/Transformers/AddNameScopeRegistration.cs

```diff
@@ -41,54 +41,99 @@ public IXamlIlAstNode Transform(XamlIlAstTransformationContext context, IXamlIlA
                         Children =
                         {
                             pa,
-                            new ScopeRegistrationNode(value)
+                            new AvaloniaNameScopeRegistrationXamlIlNode(value, context.GetAvaloniaTypes())
                         }
                     };
             }
 
+            if (!context.ParentNodes().Any()
+                && node is XamlIlValueWithManipulationNode mnode)
```

Comment:
Is there a case where we'd have just a raw root `XamlIlObjectNode` instead of a manipulation node? Why don't we just compare against the `context.RootObject` property here?

Reply:
Compiler expects to have `XamlIlValueWithManipulationNode` as the root after AST transformations are done. It splits the value and manipulation, value is used for object creation and completely skipped for `Populate`: https://github.com/kekekeks/XamlIl/blob/master/src/XamlIl/Transform/XamlIlCompiler.cs#L195

Reply:
For the compiled-binding work I've had to move this transform phase a little earlier and this is no longer true. Where is the root object changed to a `XamlIlValueWithManipulationNode`?

Reply:
`XamlIlNewObjectTransformer`. Note that XamlIlRootObjectNode only holds the type of the root object and just returns the value of a field in `XamlIlContext` runtime-support class.

Reply:
That explains it. I moved this phase to run right before that one. I ended up just splitting the root namescope part into a separate transformer that runs where this phase used to run.

Reply:
We might want to split adding `HandleRootObjectScopeNode` to a separate AST transformer since it really has to run after everything else at runtime since it calls `INameScope:Complete` which prevents further registrations.

Reply:
That’s what I ended up doing. I think it actually runs at the very end in my compiled bindings pr.